### PR TITLE
feat: Update Docker and CI

### LINKED ISSUE #2412: Add Docker image documenation and default Docker registry.
### What new feature would you like to see?

Docker container image documentation and a default registry to pull it from, such as Docker Hub?

### Which of these categories fit your request?

Other

### Just checking...

- [X] I have confirmed that my feature has not already been requested.
- [X] I have confirmed that my feature has not already been implemented in the `review` branch.

### PR DESCRIPTION
**Pull request by pythoninthegrass**
- [x] I have tested my changes against the `dev` branch (the latest developmental version), and this pull request is targeting that branch as a base
- [x] I have tested my changes on Python 3.8 or higher
- [ ] I have ensured my code is formatted using [Black](https://github.com/psf/black)

----

### Description

Original PR at #2431

> [!NOTE] 
> This PR includes recent updates from the `dev` branch. My specific contributions are focused on Docker and CI/CD improvements, as detailed in the commit "feat: update docker and CI/CD". 
>
> I did my best to rebase my feature branch against the upstream dev, but ended up coauthoring a handful of commits.
> 
> Hoping that since github says the branch can be merged automatically, that should be sufficient 🤞 
>

My key changes include:

- Updated `docker.yml` for multi-architecture builds and container registry pushes
- Refined `Dockerfile` with buildkit syntax and environment variable improvements
- Refactored `docker-compose.example.yml`
- Updated `docker-entrypoint.sh` to remove bashisms and control for word splitting
- Added `.dockerignore`, `.dive-ci`, and `.hadolint.yaml` for Docker build optimization and linting
- Updated `.gitignore` and added `.env.example`

These changes are aimed at improving the Docker-related aspects of the project and enhancing the CI/CD pipeline.

### Related issues (if applicable)

Closes #2412

### CODE REVIEW

**Comment by BabyBoySnow** [comment]
It seems that the rebase has reverted some of the features in the dev branch

**Comment by itsTheFae** [comment]
Yeah, this wont fly either.  Sorry.   
You need to make a clean copy of dev and redo your work here.  You cannot rebase master onto dev.   

Since this PR is from a branch, you should be able to force-push the clean copy and subsequent changes to that branch and reset the commits of this PR automatically.

**Comment by srhinos** [comment]
the above comments are all very valid takes.
To add on, while I understand your reasoning behind adding them, theres no reason to add the following:
- `.env.example` - we dont need a `.env` just for CI, these things should be contained to just the CI.
- `.dive-ci` - I have never heard of this and see no benefit to it
- `.hadolint.yaml` - I have never heard of this and see no benefit to it

Beyond that, I'd say cut out all the extra fluff added. No one needs a defined docker compose network for this bot. That changes nothing other than increased diff. Less is more

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    paths:
+      - 'Dockerfile*'
+      - 'pyproject.toml'
+      - 'poetry.lock'
+      - 'requirements.txt'
+      - '**.py'
+      - '**.sh'
+      - '.dockerignore'
+      - '.env.example'
+      - '.github/workflows/**'
```
Comment (srhinos) [inline .github/workflows/docker.yml:19]: what will this actually do? this seems largely pointless and should just be removed

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      - name: Build and push
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          platforms: linux/amd64,linux/arm64/v8
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          cache-from: type=registry,ref=${{ env.REGISTRY_URL }}/${{ env.REGISTRY_USER }}/${{ steps.image_name.outputs.IMAGE }}:buildcache
+          cache-to: type=registry,ref=${{ env.REGISTRY_URL }}/${{ env.REGISTRY_USER }}/${{ steps.image_name.outputs.IMAGE }}:buildcache,mode=max
```
Comment (srhinos) [inline .github/workflows/docker.yml:102]: this is all so insanely complex for a simple task such as uploading a built docker image to github's container registry.

I built this exact workflow in another repo I maintain @ [here](https://github.com/srhinos/primelooter/blob/main/.github/workflows/release.yml), feel free to use it as inspiration as this workflow is really awful.

File: Dockerfile [hunk 3]
```diff
@@ -7,22 +9,33 @@ COPY ./config sample_config
...
+    opus-dev \
+    && rm -rf /var/cache/apk/*
+
+# pip env vars
+ENV PIP_NO_CACHE_DIR=off
+ENV PIP_DISABLE_PIP_VERSION_CHECK=on
+ENV PIP_DEFAULT_TIMEOUT=100
+
+# don't generate .pyc, enable tracebacks on seg faults
+ENV PYTHONDONTWRITEBYTECODE=1
```
Comment (srhinos) [inline Dockerfile:34]: we dont want this enabled. It greatly reduces performance

File: Dockerfile [hunk 4]
```diff
@@ -7,22 +9,33 @@ COPY ./config sample_config
...
+    git \
+    libffi \
+    libsodium \
+    opus-dev \
+    && rm -rf /var/cache/apk/*
+
+# pip env vars
+ENV PIP_NO_CACHE_DIR=off
+ENV PIP_DISABLE_PIP_VERSION_CHECK=on
+ENV PIP_DEFAULT_TIMEOUT=100
```
Comment (srhinos) [inline Dockerfile:31]: these aren't needed as well. No reason to add them

File: Dockerfile [hunk 5]
```diff
@@ -1,4 +1,6 @@
-FROM python:3.8-alpine
+# syntax=docker/dockerfile:1.7.0
+
+FROM python:3.8-alpine3.20
```
Comment (srhinos) [inline Dockerfile:3]: is there any reason to deeply specify the alpine version?

File: docker-compose.example.yml [hunk 6]
```diff
@@ -1,8 +1,23 @@
...
   musicbot:
-    build: .
+    container_name: musicbot
+    build:
+      context: .
+      dockerfile: Dockerfile
+    hostname: musicbot
+    platform: ${PLATFORM:-linux/amd64}
+    environment:
+      - APP_ENV=docker
```
Comment (srhinos) [inline docker-compose.example.yml:10]: this isnt really needed unless we're using this in code somewhere.

File: docker-compose.example.yml [hunk 7]
```diff
@@ -1,8 +1,23 @@
...
       - ./data:/musicbot/data
       - ./logs:/musicbot/logs
+    networks:
+      - musicbot
+    working_dir: /musicbot
+
+networks:
+  musicbot:
+    name: musicbot
+    driver: bridge
```
Comment (srhinos) [inline docker-compose.example.yml:23]: none of this networking stuff is needed.

**Review by srhinos — CHANGES_REQUESTED** [review]
the above comments are all very valid takes.
To add on, while I understand your reasoning behind adding them, theres no reason to add the following:
- `.env.example` - we dont need a `.env` just for CI, these things should be contained to just the CI.
- `.dive-ci` - I have never heard of this and see no benefit to it
- `.hadolint.yaml` - I have never heard of this and see no benefit to it

Beyond that, I'd say cut out all the extra fluff added. No one needs a defined docker compose network for this bot. That changes nothing other than increased diff. Less is more

**Comment by pythoninthegrass** [comment]
Don't have the bandwidth for this. Take what you want or not.

glhf ✌️
________________________________

**Comment by BabyBoySnow** [comment]
@pythoninthegrass fwiw if you have a cellular device with unlimited data I recommend looking into pdanet+ an app which allows you to use your phones data on your computer. (Separate from hotspot)

**Pull request closed without merging**